wpical/src/main/native/cpp/fieldcalibration.cpp

```diff
@@ -0,0 +1,824 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#include "fieldcalibration.h"
+
+class PoseGraphError
+{
+public:
+  PoseGraphError(Pose t_ab_observed)
+      : m_t_ab_observed(std::move(t_ab_observed)) {}
+
+  template <typename T>
+  bool operator()(const T *const p_a_ptr, const T *const q_a_ptr,
+                  const T *const p_b_ptr, const T *const q_b_ptr,
+                  T *residuals_ptr) const
+  {
+    // Tag A
+    Eigen::Map<const Eigen::Matrix<T, 3, 1>> p_a(p_a_ptr);
+    Eigen::Map<const Eigen::Quaternion<T>> q_a(q_a_ptr);
+
+    // Tag B
+    Eigen::Map<const Eigen::Matrix<T, 3, 1>> p_b(p_b_ptr);
+    Eigen::Map<const Eigen::Quaternion<T>> q_b(q_b_ptr);
+
+    // Rotation between tag A to tag B
+    Eigen::Quaternion<T> q_a_inverse = q_a.conjugate();
+    Eigen::Quaternion<T> q_ab_estimated = q_a_inverse * q_b;
+
+    // Displacement between tag A and tag B in tag A's frame
+    Eigen::Matrix<T, 3, 1> p_ab_estimated = q_a_inverse * (p_b - p_a);
+
+    // Error between the orientations
+    Eigen::Quaternion<T> delta_q =
+        m_t_ab_observed.q.template cast<T>() * q_ab_estimated.conjugate();
+
+    // Residuals
+    Eigen::Map<Eigen::Matrix<T, 6, 1>> residuals(residuals_ptr);
+    residuals.template block<3, 1>(0, 0) =
+        p_ab_estimated - m_t_ab_observed.p.template cast<T>();
+    residuals.template block<3, 1>(3, 0) = T(2.0) * delta_q.vec();
+
+    return true;
+  }
+
+  static ceres::CostFunction *Create(const Pose &t_ab_observed)
+  {
+    return new ceres::AutoDiffCostFunction<PoseGraphError, 6, 3, 4, 3, 4>(
+        new PoseGraphError(t_ab_observed));
+  }
+
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+private:
+  const Pose m_t_ab_observed;
+};
+
+std::tuple<Eigen::Matrix<double, 3, 3>, Eigen::Matrix<double, 8, 1>>
+fieldcalibration::load_camera_model(std::string path)
+{
+  Eigen::Matrix<double, 3, 3> camera_matrix;
+  Eigen::Matrix<double, 8, 1> camera_distortion;
+
+  std::ifstream file(path);
+
+  nlohmann::json json_data;
+
+  try
+  {
+    json_data = nlohmann::json::parse(file);
+  }
+  catch (const nlohmann::json::parse_error &e)
+  {
+    std::cout << e.what() << std::endl;
+  }
+
+  bool isCalibdb = json_data.contains("camera");
+
+  if (!isCalibdb)
+  {
+    for (int i = 0; i < camera_matrix.rows(); i++)
+    {
+      for (int j = 0; j < camera_matrix.cols(); j++)
+      {
+        camera_matrix(i, j) =
+            json_data["camera_matrix"][(i * camera_matrix.cols()) + j];
+      }
```

Comment:
Consider replacing with a specialization of frc::from_json, like here: https://github.com/wpilibsuite/allwpilib/blob/caf424e09ca379bcf3e1807765b0b6793a648fce/apriltag/src/main/native/cpp/AprilTag.cpp#L58